### PR TITLE
Fix `recursive-protection` feature flag

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

- Closes https://github.com/apache/datafusion/issues/13766

## Rationale for this change

The recursive-protect flag was added in https://github.com/apache/datafusion/pull/13778, but when @andygrove  tried to disable it in comet, recursive was still enabled. 

I am not an expert in the crate feature mechanism, but the feature appears to get activated if a downstream crate like comet uses a crate like `datafusion` (the core crate) that passes the feature through

## What changes are included in this PR?

1. Change feature flag to enable recursive protection at the main crate level
2. rename `recursive-protection` to `recursive_protection` to be consistent with other flags

## Are these changes tested?

1. Make feature not enabled by default in sub crates
3. Add feature to the datafusion core crate
4. move another stack protection mechanism into a guarded section

Here is an example showing how I tested to verify that the feature flagging still works

```shell
$ cargo test --features=recursive-protection -p datafusion-common --lib | grep test_large_tree
    Finished `test` profile [unoptimized + debuginfo] target(s) in 0.08s
     Running unittests src/lib.rs (target/debug/deps/datafusion_common-3dd7d455d375513b)
$ cargo test -p datafusion-common --lib | grep test_large_tree
   Compiling datafusion-common v43.0.0 (/Users/andrewlamb/Software/datafusion/datafusion/common)
    Finished `test` profile [unoptimized + debuginfo] target(s) in 1.32s
     Running unittests src/lib.rs (target/debug/deps/datafusion_common-a4d89ac32a58d472)
$
````

I also verified that the MIRI tests pass on https://github.com/apache/datafusion-comet/pull/1154 with this fix

### CODE REVIEW

**Comment by alamb** [comment]
Given this experience, we really need a test for the `recursive-protection` feature flag / compiling datafusion with various feature flags. Maybe someone can figure out how to do so, following the model of arrow-rs:

https://github.com/apache/arrow-rs/blob/2c84f243b882eff69806cd7294d38bf422fdb24a/.github/workflows/arrow.yml#L63-L97

Maybe after @Omega359 's PR here we can add these tests
- https://github.com/apache/datafusion/pull/13876

File: datafusion/common/Cargo.toml [hunk 1]
```diff
@@ -36,7 +36,6 @@ name = "datafusion_common"
 path = "src/lib.rs"
 
 [features]
-default = ["recursive-protection"]
```
Comment (alamb) [inline datafusion/common/Cargo.toml:39]: having the feature enabled by default in subcrates meant I think it basically was not possible to disable when using datafusion core crate

File: datafusion/core/Cargo.toml [hunk 2]
```diff
@@ -69,6 +69,13 @@ pyarrow = ["datafusion-common/pyarrow", "parquet"]
 regex_expressions = [
     "datafusion-functions/regex_expressions",
 ]
+recusive-protection = [
```
Comment (alamb) [inline datafusion/core/Cargo.toml:72]: This adds the `recursive-protection` feature to the `datafusion` (core) crate which then activates the feature in the subcrates

File: datafusion/sql/src/query.rs [hunk 3]
```diff
@@ -62,10 +63,11 @@ impl<S: ContextProvider> SqlToRel<'_, S> {
                 // The functions called from `set_expr_to_plan()` need more than 128KB
                 // stack in debug builds as investigated in:
                 // https://github.com/apache/datafusion/pull/13310#discussion_r1836813902
-                let min_stack_size = recursive::get_minimum_stack_size();
```
Comment (alamb) [inline datafusion/sql/src/query.rs:65]: here is pretty good evidence that the feature flag is not working as expected, as this code requires `recursive` to compile but it wasn't failing the build -- I pulled the logic its own structure

**Comment by alamb** [comment]
Also before I merge this I want to rename the flag to `recursive_protection` (underscore) to be consistent with the rest

**Comment by alamb** [comment]
FYI @buraksenn and @peter-toth

**Review by xudong963 — APPROVED** [review]
LGTM.

One concern is that some users may depend on the feature, so maybe we should highlight the change at the next release log.

**Comment by alamb** [comment]
> LGTM.
> 
> One concern is that some users may depend on the feature, so maybe we should highlight the change at the next release log.

THis is a good point -- I believe 44.0.0 will be the first release that will have this feature -- 43 did not have it: https://github.com/apache/datafusion/blob/main/dev/changelog/43.0.0.md

**Review by andygrove — APPROVED** [review]
LGTM but I have not tested with Comet. Thanks @alamb

**Comment by alamb** [comment]
> LGTM but I have not tested with Comet. Thanks @alamb

I have tested so locally -- specifically I checked out the code in https://github.com/apache/datafusion-comet/pull/1154

Changed it to use a local checkout with this PR:

```diff
--- a/native/Cargo.toml
+++ b/native/Cargo.toml
@@ -39,15 +39,15 @@ arrow-buffer = { version = "53.3.0" }
 arrow-data = { version = "53.3.0" }
 arrow-schema = { version = "53.3.0" }
 parquet = { version = "53.3.0", default-features = false, features = ["experimental"] }
-datafusion = { git = "https://github.com/apache/datafusion.git", rev = "242f45f", default-features = false, features = ["unicode_expressions", "crypto_expressions"] }
-datafusion-common = { git = "https://github.com/apache/datafusion.git", rev = "242f45f", default-features = false }
-datafusion-functions = { git = "https://github.com/apache/datafusion.git", rev = "242f45f", default-features = false, features = ["crypto_expressions"] }
-datafusion-functions-nested = { git = "https://github.com/apache/datafusion.git", rev = "242f45f", default-features = false }
-datafusion-expr = { git = "https://github.com/apache/datafusion.git", rev = "242f45f", default-features = false }
-datafusion-expr-common = { git = "https://github.com/apache/datafusion.git", rev = "242f45f", default-features = false }
-datafusion-execution = { git = "https://github.com/apache/datafusion.git", rev = "242f45f", default-features = false }
-datafusion-physical-plan = { git = "https://github.com/apache/datafusion.git", rev = "242f45f", default-features = false }
-datafusion-physical-expr = { git = "https://github.com/apache/datafusion.git", rev = "242f45f", default-features = false }
+datafusion = { path = "/Users/andrewlamb/Software/datafusion/datafusion/core", default-features = false, features = ["unicode_expressions", "crypto_expressions"] }
+datafusion-common = { path = "/Users/andrewlamb/Software/datafusion/datafusion/common", default-features = false }
+datafusion-functions = { path = "/Users/andrewlamb/Software/datafusion/datafusion/functions", default-features = false, features = ["crypto_expressions"] }
+datafusion-functions-nested = { path = "/Users/andrewlamb/Software/datafusion/datafusion/functions-nested", default-features = false }
+datafusion-expr = { path = "/Users/andrewlamb/Software/datafusion/datafusion/expr", default-features = false }
+datafusion-expr-common = { path = "/Users/andrewlamb/Software/datafusion/datafusion/expr-common", default-features = false }
+datafusion-execution = { path = "/Users/andrewlamb/Software/datafusion/datafusion/execution", default-features = false }
+datafusion-physical-plan = { path = "/Users/andrewlamb/Software/datafusion/datafusion/core", default-features = false }
+datafusion-physical-expr = { path = "/Users/andrewlamb/Software/datafusion/datafusion/physical-expr", default-features = false }
 datafusion-comet-spark-expr = { path = "spark-expr", version = "0.5.0" }
 datafusion-comet-proto = { path = "proto", version = "0.5.0" }
 chrono = { version = "0.4", default-features = false, features = ["clock"] }
```
and verified that this command passes:
```shell
MIRIFLAGS="-Zmiri-disable-isolation" cargo +nightly miri test -p datafusion-comet -- test_unpack_dictionary_primitive
```

**Comment by alamb** [comment]
Thank you for the reviews @andygrove  and @xudong963

File: datafusion/sql/src/query.rs (hunk 3 above)
Comment (peter-toth) [reply datafusion/sql/src/query.rs:65]: Late LGTM. `StackGuard` is a very clever fix.

**Pull request merged**